spec/models/container_label_tag_mapping_spec.rb

```diff
@@ -178,7 +184,293 @@ def map_to_tags(mapper, model_name, labels_kv)
+        end
+
+        before do
+          FactoryBot.create(:container_label_tag_mapping, :all_entities, :label_name => "Locality", :tag => single_value_category_2.tag)
+        end
+
+        context "when mappings are targeted to different categories" do
+          it "adds tags to resource" do
+            expect(vm.tags).to be_empty
+
+            subject
+
+            expected_tags = %w[location/brno environment/accounting].map { |x| Tag.lookup_by_classification_name(x) }
+            expect(vm.reload.tags).to match_array(expected_tags)
+          end
+        end
+
+        context "when mappings are targeted to same categories" do
+          let(:externals_labels) do # provider label
+            [{:name => 'Name', :value => 'Windows'},
+             {:name => 'Location', :value => 'Production'},
+             {:name => 'Cost Center', :value => '007'}]
+          end
+
+          before do
+            FactoryBot.create(:container_label_tag_mapping, :all_entities, :label_name => "Location", :tag => single_value_category.tag)
+            FactoryBot.create(:container_label_tag_mapping, :all_entities, :label_name => "Cost Center", :tag => single_value_category.tag)
+          end
+
+          it "only applies one of the label values" do
+            expect(vm.tags).to be_empty
+
+            subject
+            expect(vm.reload.tags.count).to eq(1)
+
+            # Only first tag is assigned to Vm and this tag is selected according to
+            # alphanumeric order of labels's names
+            expect(vm.reload.tags.first.name).to eq("/managed/environment/007")
+          end
+        end
+      end
+    end
+
+    context "with mapping with multi-value category" do
+      let!(:multi_value_category_2) { FactoryBot.create(:classification_cost_center_with_tags) }
+
+      let(:externals_labels) do # provider label
+        [{:name => 'Name', :value => 'Accounting'}]
+      end
+
+      before do
+        FactoryBot.create(:container_label_tag_mapping, :all_entities, :label_name => "Name", :tag => multi_value_category.tag)
+      end
+
+      it "adds a tag to the resource with the value of the external label" do
+        expect(vm.tags).to be_empty
+
+        subject
+
+        expected_tags = [Tag.lookup_by_classification_name('department/accounting')]
+        expect(vm.reload.tags).to eq(expected_tags)
+      end
+
+      it "reassigns to the tag specified in the mapping with same category as a pre-existing tag" do
+        expect(vm.tags).to be_empty
+
+        existing_tag_on_vm = Tag.lookup_by_classification_name('department/hr')
+        vm.tag_add(existing_tag_on_vm.name, :ns => '')
+        expect(vm.reload.tags).to eq([existing_tag_on_vm])
+
+        subject
+
+        expect(vm.reload.tags).to match_array([Tag.lookup_by_classification_name('department/accounting')])
+      end
+
+      it "add tag with a category that is different from the target category from mapping " do
```

Comment:
This wording is difficult to follow. I'll see if I can come up with something better